crates/bevy_ecs/src/system/system_registry.rs

```diff
@@ -500,6 +500,16 @@ impl World {
 
     /// Runs a cached system, registering it if necessary.
     ///
+    /// # Type Inference Note
+    /// If the system returns `()`, you may need to explicitly constrain the output
+    /// type for error handling:
+    ///
+    /// ```rust
+    /// () = world.run_system_cached(my_system)?;
+    /// ```
+    ///
+    /// Without this, Rust may fail to infer the system’s output type and produce
+    /// a `IntoResult<!>` inference error.
```

Comment:
Huh, I thought the error was "cannot infer type of the type parameter `O` declared on the method `run_system_cached`".  

... Oh, I see, `world.run_system_cached(my_system).unwrap()` gives the helpful "cannot infer type" message, but `world.run_system_cached(my_system)?` gives the unhelpful "the trait `function_system::IntoResult<!>` is not implemented for `core::result::Result<(), bevy_error::BevyError>`".  I wonder why that is.  ... Huh, and if `my_system` returns `!`, then `unwrap()` still gives "cannot infer type", but `?` actually compiles.  So the compiler really assumes it's `!` for some reason.  

---

Anyway, it might be helpful to give a little more detail on the inference error, like:

```suggestion
    /// an error either of the form
    ///
    /// ```text
    /// the trait `IntoResult<!>` is not implemented for `Result<(), BevyError>`
    /// ```
    ///
    /// or
    ///
    /// ```text
    /// cannot infer type of the type parameter `O` declared on the method `run_system_cached`
    /// ```
```